kore/src/Kore/Builtin/Bool.hs

```diff
@@ -163,48 +165,63 @@ builtinFunctions =
     xor a b = (a && not b) || (not a && b)
     implies a b = not a || b
 
+data UnifyBool = UnifyBool
+    { bool1, bool2 :: InternalBool
+    }
+
+matchBools ::
+    TermLike RewritingVariableName ->
+    TermLike RewritingVariableName ->
+    Maybe UnifyBool
+matchBools first second
+    | InternalBool_ bool1 <- first
+      , InternalBool_ bool2 <- second =
+        Just $ UnifyBool bool1 bool2
+    | otherwise = Nothing
+{-# INLINE matchBools #-}
+
 -- | Unification of @BOOL.Bool@ values.
 unifyBool ::
-    forall variable unifier.
-    InternalVariable variable =>
+    forall unifier.
     MonadUnify unifier =>
-    TermLike variable ->
-    TermLike variable ->
-    MaybeT unifier (Pattern variable)
-unifyBool a b =
-    worker a b <|> worker b a
+    TermLike RewritingVariableName ->
+    TermLike RewritingVariableName ->
+    UnifyBool ->
+    unifier (Pattern RewritingVariableName)
+unifyBool termLike1 termLike2 unifyData =
+    if bool1 == bool2
+        then return (Pattern.fromTermLike termLike1)
+        else
+            Unify.explainAndReturnBottom
+                "different Bool domain values"
+                termLike1
+                termLike2
   where
-    worker termLike1 termLike2
-        | Just value1 <- matchBool termLike1
-          , Just value2 <- matchBool termLike2 =
-            lift $
-                if value1 == value2
-                    then return (Pattern.fromTermLike termLike1)
-                    else
-                        Unify.explainAndReturnBottom
-                            "different Bool domain values"
-                            termLike1
-                            termLike2
-    worker _ _ = empty
+    UnifyBool{bool1, bool2} = unifyData
+
+matchUnifyBoolAnd ::
+    TermLike RewritingVariableName ->
+    TermLike RewritingVariableName ->
+    Maybe (BoolAnd (TermLike RewritingVariableName))
+matchUnifyBoolAnd first second
```

Comment:
Where is the `UnifyBoolAnd` type?

Reply:
There is none, here I am following the convention that if the match is returning only a single datum that we just return that without wrapping.